llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -5083,7 +5083,18 @@ SDValue DAGCombiner::visitAVG(SDNode *N) {
   if (N0 == N1 && Level >= AfterLegalizeTypes)
     return N0;
 
-  // TODO If we use avg for scalars anywhere, we can add (avgfl x, 0) -> x >> 1
+  // Fold (avg x, 0) -> x >> 1
+  if (isNullOrNullSplat(N0))
```

Comment:
Don't we canonicalize constants to RHS earlier?

Reply:
Can't you just remove the `isVector` check around the existing code on line 5067?

Reply:
Not the same behavior. isNullOrNullSplat returns false if there are any undefs in the vector.

Reply:
Then replace that code with your preferred version.

Reply:
Also if you are intentionally improving the handling of vectors with undefs then you should add tests for that.

Reply:
isConstantSplatVectorAllZeros is for vectors. isNullOrNullSplat is for constants and vectors.

Reply:
If you pass allow undefs to `isNullOrNullSplat`, can we remove the `ISD::isConstantSplatVectorAllZeros` version of the code?